### DISCUSSION TITLE
What about Iif()? - shouldn't we do something about this?

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#644] [Ideas]
Let's face it - Iif() is a bit of a mess - especially for those of us who do Access VBA or use EVAL() much.

Situationally, IIF() either short-circuits, or it doesn't, and the fact that the same syntax can behave differently in seemingly slightly different use cases is both confusing and (potentially, at least) bug-inducing - and that pisses all over the ideal of code clarity.
(see: https://nolongerset.com/iif-function-vs-statement/ and https://nolongerset.com/ternary-operator-iif/ for more background if needed.)

Now, I get that we need to be 100% backwards compatible, but can't we do something here to make this particular situation better?

Perhaps an OPTION ENHANCED type of thing where a new IF() will short-circuit, and IIF() won't? or the reverse might be a better (if perhaps less intuitive) idea, since we can't exactly alter the implementation behavior of iif() within the Jet Expression service evaluator.

As it bothers Mike, it also bothers me.

...or just call me a pedant too.

**Comment by WaynePhillipsEA** [comment]
I agree.  tB offers an If() short-circuiting function, and leaves the old Iif() version intact for VBx compatibility.

**Reply by mburns08109** [reply]
But wayne, that approach (while inarguably an improvement) leaves the fundamental problem intact.
iif() is still a mess in an of itself, and adding if() alone, doesn't change that at all.

VBA.iif() needs to be chanagable* to match what the JET Expression service iif() does. Otherwise the inherent problem will remain.

*Note that I said "changable", rather than "changed".

**Reply by Kr00l** [reply]
Changing behaviors of an function via an "enhancement option" is a two sided sword.
It will break compatibility at the end..

**Reply by bclothier** [reply]
That's what refactoring is for. Trust me, you really really don't want a quirk mode. You really want to refactor your code.

**Reply by mburns08109** [reply]
but the developer would have control over that - and for new code, that would be OK IMO.

**Reply by WaynePhillipsEA** [reply]
I do understand that, but we have to support `Iif` exactly as it is implemented in VBx, there's no getting around that as backwards compatibility is the absolute top priority of this project.  

The general consensus, from what I've picked up from early discussions, is that we don't want to add loads of options/switches to change fundamental behaviour such as this.  Therefore, having the additional `If` function is likely the best that we can manage.

**Reply by mburns08109** [reply]
> That's what refactoring is for. Trust me, you really really don't want a quirk mode. You really want to refactor your code.

But ben, haven't you been arguing FOR a Quirk Mode in tB all along here already? this is sort of the opposite-case of that thinking.

**Reply by bclothier** [reply]
If I was, I did a poor job of it as I've never been in favor for it...

**Reply by mburns08109** [reply]
Ok, Mike, I gave it a try, your turn to see if you can convince them, I guess.   :-)

**Reply by mwolfe02** [reply]
Mark, I'm not in favor of changing (even optionally) the behavior of the VBA.IIf() function.  We can't change the default behavior for backward-compatibility reasons and adding switches to change behavior--especially something as fundamental as changing a library function into an operator--strikes me as a very bad idea.

Truth be told, I'm not that keen on introducing a language operator (i.e., `If()`) that looks like a function.  I think it would be better if all the language-level operators were implemented with symbols.  **_However_**, that ship has sailed.  The `Mod()` operator took care of that, so there's already prior art in VBx for using keywords for language operators.  There's also prior art for the `If()` ternary operator in VB.Net.  

With those two things in mind, I don't see any other obviously better alternative than leaving `IIf()` _the function_ alone and implementing the _ternary operator_ using the `If()` keyword.

**Reply by bclothier** [reply]
> The `Mod()` operator took care of that,

I just want to point out that one thing that the `Mod` operator as well as `And` or `Or` has going for them is that they _don't_  use the parentheses --- it's `2 = x Mod 5`, not `2 = Mod(x, 5)`. So they still have that aura of an operator. The `If()`, OTOH definitely looks far too functionish.

It would have been a bit more operator-ish to see: `2 = If 1 = 1 Then True Else 1 / 0`, but I suspect this opens a huge cans of worms. TBH reading it, it feels more weird than seeing the `If()` which clicks better. As you said, the ship already has sailed but I wanted to call attention to the difference between the `If()` and `Mod` for future discussions.

**Reply by VBFNiya** [reply]
>The If(), OTOH definitely looks far too functionish.

There is precedent for this. I believe some of you refer to it as "prior art". sizeof in C is classified as an operator. VB.Net has IIf and If. IIf is an actual function and If is an operator. Given all this information it seems to me that the way functional style operators are differentiated from actual functions lie in how they are implemented. If it's implemented in the compiler as in the case of sizeof in C and If in VB.Net then it's an operator. If it's implemented in a library, then it's a function.

My personal opinion is that things like sizeof and If should not be classified as operators but that ship has sailed a lifetime ago. These ideas are already set in stone.

**Reply by wqweto** [reply]
In C/C++ something like `sizeof int` and `sizeof(int)` are valid expressions i.e. the operator does not require parentheses the same way both `return 0` and `return(0)` are valid `return` statements.

The idea is that in `sizeof <expr>` you can have extra parens in `<expr>` so it looks like a function call but is not (the same way `return(0)` has extra parens).

This is very much the way beginners in VBx write `MySub (42)` and complain when `MySub (42, true)` fails to compile :-)) (Not in TB already.)

**Reply by vbBaka** [reply]
well, we already have If() now. and its good enough. if u can deal with << >> and other operators Im sure u get used to if() as well.

**Reply by VBFNiya** [reply]
>In C/C++ something like sizeof int and sizeof(int) are valid expressions i.e. the operator does not require parentheses the same way both return 0 and return(0) are valid return statements.

That actually makes sense when you think about it. I mean think of any other operator, say the + operator. You can do 1+3 or 1+(3). Or if you use a similar prefix operator like +=, you can do stuff like i+=(3). The point is, it's actually consistent with what you'd expect from an operator so I guess it perhaps wasn't such a bad idea to classify it as such.

**Reply by wqweto** [reply]
> well, we already have If() now. and its good enough. if u can deal with << >> and other operators Im sure u get used to if() as well.

No, we will sure insist on bringing the ternary operator from C/C++ because I just cannot live without writing `condition ? 42 : 0` in the morning :-))

**Reply by vbBaka** [reply]
looks beautiful
msg = (num === 4) ? "Correct!" : "Incorrect!";
a most have.

**Reply by VBFNiya** [reply]
> No, we will sure insist on bringing the ternary operator from C/C++ because I just cannot live without writing condition ? 42 : 0 in the morning :-))

This actually might not be such a bad idea.

**Reply by Greedquest** [reply]
@bclothier 
> It would have been a bit more operator-ish to see: `x = If 1 = 1 Then True Else 1 / 0`, but I suspect this opens a huge cans of worms.

I think `x = True If 1=1 Else 1/0` would be the best ternary operator version of this (a bit more basic than the `? :` stuff)

**Comment by EduardoVB** [comment]
What about `IIF2` (or some other name)?
Or another possibility, IDK if it would be possible to add a final flag parameter to `IIF` for short-circuiting?

**Reply by Kr00l** [reply]
> What about `IIF2` (or some other name)? Or another possibility, IDK if it would be possible to add a final flag parameter to `IIF` for short-circuiting?

Why adding new `IIF2` when you can use the existing `IF`

**Reply by bclothier** [reply]
Just in case it wasn't already clear from Wayne's earlier comment.... You already have it so this discussion is moot.

```
    Public Function IfExample() As Boolean
        Return If(1 = 1, True, 1 / 0)
    End Function
```

**Reply by EduardoVB** [reply]
OK, sorry, I didn't remember that (now I recall).
